### PR TITLE
xds: Required Router Filter for both Client and Server side

### PR DESCRIPTION
**Pull request by zasweq**
This PR adds a requirement for Router Filter for both Client and Server Side. This is done through adding an API you can call to see if an HTTP Filter is terminal.

RELEASE NOTES: none

### CODE REVIEW

File: xds/internal/httpfilter/httpfilter.go [hunk 1]
```diff
@@ -50,6 +50,11 @@ type Filter interface {
 	// not accept a custom type.  The resulting FilterConfig will later be
 	// passed to Build.
 	ParseFilterConfigOverride(proto.Message) (FilterConfig, error)
+	// IsTerminal returns whether this Filter is terminal or not (i.e. it must
+	// be last filter in the filter chain). Currently, grpc only supports the
+	// router filter as a terminal filter, as per A39, although this may change
```
Comment (easwars) [inline xds/internal/httpfilter/httpfilter.go:55]: Does this comment saying "router filter is the only terminal filter" need to be part of the interface definition?

Also, isnt HCM the terminal filter on the server?

**Review by menghanl — CHANGES_REQUESTED** [review]

File: xds/internal/xdsclient/xds.go [hunk 2]
```diff
@@ -244,6 +244,14 @@ func processHTTPFilters(filters []*v3httppb.HttpFilter, server bool) ([]HTTPFilt
 		// Save name/config
 		ret = append(ret, HTTPFilter{Name: name, Filter: httpFilter, Config: config})
 	}
+	// "Validation will fail if a terminal filter is not the last filter in the
```
[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/xds.go:247]: How does this not cover that: if !ret[len(ret)-1].Filter.IsTerminal() {
		return nil, fmt.Errorf("http filter %q is not a terminal filter", ret[len(ret)-1].Name)
	}

[the remark this answers is not in the document]
Comment (menghanl) [reply xds/internal/xdsclient/xds.go:247]: Validation will fail
- if a terminal filter is not the last filter in the chain or
  - --> terminal must be last, this is not covered
- if a non-terminal filter is the last filter in the chain
  - --> last must be terminal, this is covered

**Review by menghanl — CHANGES_REQUESTED** [review]

File: xds/internal/xdsclient/filter_chain_test.go [hunk 3]
```diff
@@ -77,6 +81,9 @@ var (
 		Name:       "serverOnlyCustomFilter2",
 		ConfigType: &v3httppb.HttpFilter_TypedConfig{TypedConfig: serverOnlyCustomFilterConfig},
 	}
+	routerFilterPb   = e2e.RouterHTTPFilter
+	routerFilter     = HTTPFilter{Name: "router", Filter: httpfilter.Get(router.TypeURL), Config: router.ConfigForTesting{}}
```
[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/filter_chain_test.go:85]: Ah, wow what a beautiful solution. Thanks for the suggestion. I knew there was a better way than this Alias.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments everyone :)!

**Pull request merged**